notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,458 @@
+    "    y_group_0_5 = sum((group[\"score\"] > threshold) & (group[\"label\"] == 1)) / n_group_1[group_name]\n",
+    "    print(\"{}:\".format(group_name))\n",
+    "    print(\"    P[Ŷ = 1 | Y = 0] = {}\".format(x_group_0_5))\n",
+    "    print(\"    P[Ŷ = 1 | Y = 1] = {}\".format(y_group_0_5))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that it never makes sense to have $x>y$ because in that case you're better off flipping labels, i.e. completely turning around the meaning of the scores. Our method automatically does that unless specified otherwise.\n",
+    "\n",
+    "## Interpolated Predictions and Probabilistic Classifiers\n",
+    "\n",
+    "This way you end up with a set of points above the diagonal line connecting (0,0) and (1,1). We calculate the convex hull based on that, because we can reach any point in between two known thresholding points by interpolation. An interpolation could be $p_0 (x_0, y_0) + p_1 (x_1, y_1)$. For our post processing algorithm that would mean that we use the rule defined by $(x_0, y_0, \\text{operation}_0)$ $\\quad p_0$ percent of the time, and the rule defined by $(x_1, y_1, \\text{operation}_1)$ $\\quad p_1$ percent of the time, thus resulting in a probabilistic classifier. Depending on the data certain fairness objectives can only be accomplished with probabilistic classifiers. However, not every use case lends itself to probabilistic classifiers, since it could mean that two people with identical features are classified differently.\n",
+    "\n",
+    "## Finding the Equalized Odds solution\n",
+    "\n",
+    "In this specific instance we don't need to do any extra work to get the convex hull, since there's only one point above the diagonal line."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "\n",
+    "for group_name, group in data_grouped_by_attribute:\n",
+    "    plt.plot(roc_points[group_name].x, roc_points[group_name].y)\n",
+    "plt.xlabel(\"$P\\ [\\ \\\\hat{Y}=1\\ |\\ Y=0\\ ]$\")\n",
+    "plt.ylabel(\"$P\\ [\\ \\\\hat{Y}=1\\ |\\ Y=1\\ ]$\")\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the Equalized Odds case, we need to enforce the presented probabilities to match, which is equivalent to finding the minimum error overlap. The error in the chart is smallest in the top left corner. This is done as part of the `fit` step above, and we'll repeat it here for completeness. The yellow area is the overlap between the areas under the curve that are reachable with interpolation for both groups. Of course, this works for more than two groups as well. The result is that we have interpolated solutions for each group, i.e. every prediction is calculated as the weighted result of two threshold rules."
```

Comment:
"to enforce the presented probabilities to match" …. doesn't quite make sense. Consider rephrasing.

Reply:
It is actually exactly what we do there. We match the probabilities. Because that's how you satisfy EO.

Reply:
slightly rephrased to hopefully be specific enough